### PR TITLE
Improvements in `matrix::gather`: test coverage, compilation errors, performance

### PR DESCRIPTION
**Pull request by Nyrio**
In order to deprecate `copy_selected` from `ann_utils.cuh`, I wanted to make sure that the performance of `matrix::gather` was on par. But in the process I discovered that:

- Map transforms and conditional copy were not tested at all.
- In fact, most of the API in `gather.cuh` wasn't covered in tests and some of the functions didn't even compile.
- The same type `MatrixIteratorT` was used for the input and output iterators, which made it impossible to take advantage of custom iterators, as is needed in `kmeans_balanced` to convert the dataset from `T` to `float` and gather in a single step.
- The performance was really poor when `D` is small because the kernel assigns one block per row (so a block could be working on only 2 or 3 elements...)

This PR addresses all the aforementioned issues.

### CODE REVIEW

**Comment by Nyrio** [comment]
Below is a comparison of before/after performance. The speedups range from 1x to 7x.

2023-01-10_gather_if_perf

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Louis, thanks for the PR, it looks good! While we are here, could you improve a little bit the docstring of gather, I have left a few questions.

File: cpp/include/raft/matrix/gather.cuh [hunk 1]
```diff
@@ -30,10 +31,13 @@ namespace raft::matrix {
 /**
  * @brief  gather copies rows from a source matrix into a destination matrix according to a map.
```
[the remark this answers is not in the document]
Comment (Nyrio) [reply cpp/include/raft/matrix/gather.cuh]: I've improved the docs, let me know if you have further comments or are happy with the change. I also removed redundant overloads while I was there.

**Comment by Nyrio** [comment]
I'm not sure what's up with CI but it doesn't seem to be a compiler nor test error.

**Review by tfeher — APPROVED** [review]
Thanks Louis for updating the doc. The PR looks good to me.

**Comment by cjnolet** [comment]
@Nyrio the changes look good here. Before we merge this, can you verify that the changes to the template params don't break cuml downstream?

**Comment by Nyrio** [comment]
@cjnolet There is one call to `gather_if` in cuML and it shouldn't break (actually, the only thing that can potentially break with my change is the mdspan overload if all the template args are explicitly specified).

**Review by cjnolet — APPROVED** [review]
LGTM

**Pull request merged**